server/utils.go

```diff
@@ -0,0 +1,183 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/larkox/mattermost-plugin-badges/badgesmodel"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+func areRolesAllowed(userRoles []string, allowedRoles map[string]bool) bool {
+	for ar, b := range allowedRoles {
+		if !b {
+			continue
+		}
+		for _, ur := range userRoles {
+			if ar == ur {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func canGrantBadge(user model.User, badge badgesmodel.Badge, badgeType badgesmodel.BadgeTypeDefinition) bool {
+	if user.IsSystemAdmin() {
+		return true
+	}
+
+	if badgeType.CreatedBy == user.Id {
+		return true
+	}
+
+	if badge.CreatedBy == user.Id {
+		return true
+	}
+
+	blocked := badgeType.CanGrant.BlockList[user.Id]
+	if blocked {
+		return false
+	}
+
+	if areRolesAllowed(user.GetRoles(), badgeType.CanGrant.Roles) {
+		return true
+	}
+
+	allowed := badgeType.CanGrant.AllowList[user.Id]
+	if allowed {
+		return true
+	}
+
+	return badgeType.CanGrant.Everyone
+}
+
+func canCreateBadge(user model.User, badgeType badgesmodel.BadgeTypeDefinition) bool {
+	if user.IsSystemAdmin() {
+		return true
+	}
+
+	if badgeType.CreatedBy == user.Id {
+		return true
+	}
+
+	blocked := badgeType.CanCreate.BlockList[user.Id]
+	if blocked {
+		return false
+	}
+
+	if areRolesAllowed(user.GetRoles(), badgeType.CanCreate.Roles) {
+		return true
+	}
+
+	allowed := badgeType.CanCreate.AllowList[user.Id]
+	if allowed {
+		return true
+	}
+
+	return badgeType.CanCreate.Everyone
+}
+
+func canEditType(user model.User, badgeType badgesmodel.BadgeTypeDefinition) bool {
```

Comment:
Looks like it is not needed the `badgeType`, it is in preparation for future implementations?

Reply:
Yes. Right now only sysadmin can edit/create types, but eventually we may want to add more permissions that are specific for each type.

Reply:
Again, it is just a suggestion, but I would apply [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) here. If it is not needed now, I would remove it and add it when is needed, specially because is not a public API.